File: srclab-common/src/main/java/xyz/srclab/common/base/Sort.java
package xyz.srclab.common.base;

import java.util.Comparator;

/**
 * @author sunqian
 */
public class Sort {

    public static final Comparator<?> SELF_COMPARABLE_COMPARATOR = (o1, o2) ->
            ((Comparable<?>) o1).compareTo(Cast.as(o2));

    public static <T> Comparator<T> selfComparableComparator() {
        return Cast.as(SELF_COMPARABLE_COMPARATOR);
    }
}

File: srclab-common/src/main/kotlin/xyz/srclab/common/base/Sort.kt
@file:JvmName("Sort")
@file:JvmMultifileClass

package xyz.srclab.common.base

import java.util.*

fun <T> selfComparableComparator(): Comparator<T> {
    return SELF_COMPARABLE_COMPARATOR.asAny()
}

private val SELF_COMPARABLE_COMPARATOR: Comparator<*> by lazy {
    Comparator { o1: Any?, o2: Any? ->
        val c1: Comparable<Any?> = o1.asAny()
        val c2: Comparable<Any?> = o2.asAny()
        c1.compareTo(c2)
    }
}